### ISSUE TITLE
Challenges in Replicating Middlebury Dataset Results

### ISSUE THREAD

**Issue by UESTCglasgow** [NVlabs/FoundationStereo#149]
I evaluated the model on the Middlebury2014 (F and Q, training set) and ETH3D (test set) datasets without fine-tuning, and found that the performance was not satisfactory. Here are the results of my reproduction:

| dataset | BP-0.5 | BP-1 | BP-2 | EPE |
|--------|--------|--------|--------|--------|
| middlebury2014-F     | 9.46     | 2.17   | 0.79   | 0.33  |
| middlebury2014-F reproduction   | 35.1     | 12.0   | 8.82   | 8.43  |
| ETH3D | 1.26 | 0.26 | 0.08 | 0.09 |
| ETH3D reproduction | 7.83| 0.46| 0.09 | 0.27 |

The results labeled with 'reproduction' represent my own implementation and evaluation of the model, while the ones without are the results listed on the official leaderboard for the FoundationStereo model. It appears there is a significant discrepancy between the two sets of results, and I'm reaching out to you in the hope that you might be able to provide some insights or guidance on the factors that could account for the differences.

I use `disp2pfm.cpp` in [MiddEval3-SDK-1.6](https://vision.middlebury.edu/stereo/submit3/zip/MiddEval3-SDK-1.6.zip) (from https://vision.middlebury.edu/stereo/submit3) to convert grayscale `.png` to `.pfm` for both Middlebury2014 and ETH3D. The `.pfm` files were submitted to the website in correct zip format.

inference code is:
``` python
import os,sys
# os.environ["HF_ENDPOINT"] = 'https://hf-mirror.com'
code_dir = os.path.dirname(os.path.realpath(__file__))
sys.path.append(f'{code_dir}/../')
import glob
import logging
import imageio
import cv2
import torch
import numpy as np
from omegaconf import OmegaConf
from core.utils.utils import InputPadder
from core.foundation_stereo import *
from Utils import *
import time
import subprocess


def evaluate_model(args):

    os.makedirs(args.middlebury_out_dir, exist_ok=True)

    # 加载模型和配置
    ckpt_dir = args.ckpt_dir
    cfg = OmegaConf.load(f'{os.path.dirname(ckpt_dir)}/cfg.yaml')
    if 'vit_size' not in cfg:
        cfg['vit_size'] = 'vitl'
    for k in args.__dict__:
        cfg[k] = args.__dict__[k]
    args = OmegaConf.create(cfg)
    logging.info(f"args:\n{args}")
    logging.info(f"Using pretrained model from {ckpt_dir}")

    model = FoundationStereo(args)
    
    ckpt = torch.load(ckpt_dir)
    logging.info(f"ckpt global_step:{ckpt['global_step']}, epoch:{ckpt['epoch']}")
    model.load_state_dict(ckpt['model'])
    model.cuda()
    model.eval()
    scale = args.scale
    assert scale <= 1, "scale must be <=1"

    eth_output = args.eth3d_out_dir

    for scene in glob.glob(args.middlebury_data_dir + "/*"):
        scene_name = os.path.basename(scene)

        img0 = imageio.v2.imread(os.path.join(scene, 'im0.png'))
        img1 = imageio.v2.imread(os.path.join(scene, 'im1.png'))

        img0 = cv2.resize(img0, fx=scale, fy=scale, dsize=None)
        img1 = cv2.resize(img1, fx=scale, fy=scale, dsize=None)
        H, W = img0.shape[:2]
        img0_ori = img0.copy()
        img0 = torch.as_tensor(img0).cuda().float()[None].permute(0, 3, 1, 2)
        img1 = torch.as_tensor(img1).cuda().float()[None].permute(0, 3, 1, 2)

        padder = InputPadder(img0.shape, divis_by=32, force_square=False)
        img0, img1 = padder.pad(img0, img1)
        
        start_time = time.time()
        with torch.amp.autocast("cuda"):
            if not args.hiera:
                disp_pred = model.forward(img0, img1, iters=args.valid_iters, test_mode=True)
            else:
                disp_pred = model.run_hierachical(img0, img1, iters=args.valid_iters, test_mode=True, small_ratio=0.5)
        end_time = time.time()

        disp_pred = padder.unpad(disp_pred.float())
        disp_pred = disp_pred.data.cpu().numpy().reshape(H, W)

        # Middlebury2014
        # 保存视差图为 PNG 格式
        output_dir = os.path.join(args.middlebury_out_dir, scene_name)
        os.makedirs(output_dir, exist_ok=True)
        disp_pred_png_path = os.path.join(output_dir, f'disp0FSD.png')
        cv2.imwrite(disp_pred_png_path, disp_pred)
        # 调用 disp2pfm 工具将 PNG 转换为 PFM
        disp2pfm_path = args.disp2pfm_dir
        disp_pred_pfm_path = os.path.join(output_dir, f'disp0FSD.pfm')
        disp2pfm_cmd = f"{disp2pfm_path} {disp_pred_png_path} {disp_pred_pfm_path} 1 1"
        subprocess.run(disp2pfm_cmd, shell=True, check=True)
        # 保存时间文件
        time_txt_path = os.path.join(output_dir, f'timeFSD.txt')
        with open(time_txt_path, 'w') as f:
            f.write(f'runtime {end_time - start_time:.6f}')
        logging.info(f"Output saved to {output_dir}")

        # # ETH3D 
        # # 保存视差图为 PNG 格式
        # output_dir = eth_output
        # os.makedirs(output_dir, exist_ok=True)
        # disp_pred_png_path = os.path.join(output_dir, f'{scene_name}.png')
        # cv2.imwrite(disp_pred_png_path, disp_pred)
        # # 调用 disp2pfm 工具将 PNG 转换为 PFM
        # disp2pfm_path = args.disp2pfm_dir
        # disp_pred_pfm_path = os.path.join(output_dir, f'{scene_name}.pfm')
        # disp2pfm_cmd = f"{disp2pfm_path} {disp_pred_png_path} {disp_pred_pfm_path} 1 1"
        # subprocess.run(disp2pfm_cmd, shell=True, check=True)
        # # os.remove(disp_pred_png_path)
        # # 保存时间文件
        # time_txt_path = os.path.join(output_dir, f'{scene_name}.txt')
        # with open(time_txt_path, 'w') as f:
        #     f.write(f'runtime {end_time - start_time:.6f}')
        # logging.info(f"Output saved to {output_dir}")

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument('--ckpt_dir', default=f'{code_dir}/pretrained_models/23-51-11/model_best_bp2.pth', type=str, help='pretrained model path')
    parser.add_argument('--eth3d_data_dir', default='/workspace/MiddEval3/trainingF', type=str)
    parser.add_argument('--middlebury_data_dir', default='/workspace/MiddEval3/trainingF', type=str)
    parser.add_argument('--middlebury_out_dir', default='/workspace/MiddEval3/submit/trainingF', type=str)
    parser.add_argument('--eth3d_out_dir', default='/workspace/ETH3D/submit/two_view_training', type=str)
    parser.add_argument('--disp2pfm_dir', default="/workspace/MiddEval3/code/disp2pfm", type=str)

    # parser.add_argument('--out_png', default='/workspace/MiddEval3/trainingF', type=str)
    parser.add_argument('--scale', default=1, type=float)
    parser.add_argument('--hiera', default=0, type=int)
    parser.add_argument('--valid_iters', type=int, default=32)
    args = parser.parse_args()

    evaluate_model(args)
```

MidEval-SDK: disp2pfm.cpp is:
```cpp
// disp2pfm.cpp 

// converts pgm/png disparities to pfm, optionally dividing by dispfact (use 2 for half-size, etc.)
// if mapzero=1, 0's are considered unknown disparities and mapped to infinity

// DS 10/3/2013

static const char *usage = "\n  usage: %s disp.pgm disp.pfm [dispfact=1 [mapzero=0]]\n";

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "imageLib.h"

int verbose = 1;

// convert gray disparity image into a float image
// scale values by dispfact and convert 0's (unk values) into inf's 
void disp2floatImg(CByteImage img, CFloatImage &fimg, int dispfact, int mapzero)
{
    CShape sh = img.Shape();
    int width = sh.width, height = sh.height;
    sh.nBands = 1; 
    fimg.ReAllocate(sh);

    float s = 1.0 / dispfact;
    for (int y = 0; y < height; y++) {
	for (int x = 0; x < width; x++) {
	    int v = img.Pixel(x, y, 0);
	    float f = s * v;
	    if (v == 0 && mapzero)
		f = INFINITY;
	    fimg.Pixel(x, y, 0) = f;
	}
    }
}

int main(int argc, char *argv[])
{
    try {
	int requiredargs = 2;
	int optionalargs = 2;
	if (argc >= requiredargs + 1 && argc <= requiredargs + optionalargs + 1) {
	    int argn = 1;
	    char *dispname = argv[argn++];
	    char *fdispname = argv[argn++];
	    int dispfact = 1;
	    if (argc > argn)
		dispfact = atoi(argv[argn++]);
	    int mapzero = 0;
	    if (argc > argn)
		mapzero = atoi(argv[argn++]);
      
	    CByteImage disp;
	    CFloatImage fdisp;
		CFloatImage cdisp;
	    ReadImageVerb(disp, dispname, verbose);

	    disp2floatImg(disp, fdisp, dispfact, mapzero);

	    WriteImageVerb(fdisp, fdispname, verbose);
		// printf("KB: %f\n", fdisp.Pixel(502, 502, 0));
		ReadImageVerb(cdisp, fdispname, verbose);
		// printf("KOBE: disp=%f, name=%s\n", cdisp.Pixel(502, 502, 0), fdispname);
	} else
	    throw CError(usage, argv[0]);
    }
    catch (CError &err) {
	fprintf(stderr, err.message);
	fprintf(stderr, "\n");
	return -1;
    }
  
    return 0;
}

```

**Comment by XushengJie114514** [comment]
same problem bro

**Comment by shengnian6** [comment]
11-33-40 may be better but still low score

**Comment by wenbowen123** [comment]
> ones without are the results listed on the official leaderboard for the FoundationStereo model. 

Hi, the methods on official leaderboard are in the per-domain finetuning setup, meaning the training split was used for finetuning. The released ckpt corresponds to the zero-shot setup. Please refer to table 2 for the zero-shot results.

**Comment by UESTCglasgow** [comment]
> > ones without are the results listed on the official leaderboard for the FoundationStereo model.
> 
> Hi, the methods on official leaderboard are in the per-domain finetuning setup, meaning the training split was used for finetuning. The released ckpt corresponds to the zero-shot setup. Please refer to table 2 for the zero-shot results.

Thank you very much for your clarification. I appreciate your guidance in directing me to Table 2 for the zero-shot results. 

Upon reviewing Table 2, I noticed that the zero-shot results are indeed provided, but they appear to be limited in scope. Specifically, for the Middlebury dataset, only the BP-2 metric is reported, while for the ETH3D dataset, only the BP-1 metric is available. Additionally, I observed that Table 4 contains a more comprehensive set of zero-shot results for ETH3D. However, I found some inconsistencies between the BP-1 values reported in Table 2 and Table 4 for ETH3D, which has caused some confusion on my part.

Furthermore, I would like to kindly ask whether the released checkpoint corresponds to the model referred to as "trained only on Scene Flow" in Table 2, or the "foundation model" presented in Table 4. Clarifying this would greatly assist me in aligning my evaluation setup with the correct baseline.

Thank you again for your time and assistance. I look forward to your insights.

**Comment by wenbowen123** [comment]
> However, I found some inconsistencies between the BP-1 values reported in Table 2 and Table 4 for ETH3D, which has caused some confusion on my part.

Table 2 is the zero-shot evaluation on the training split of ETH3D since that's the only split where GT depth has been provided. This is a common evaluation protocol for zero-shot setup. Table 4 was the results on test split of ETH3D submitted to leaderboard.

The released ckpt is the foundation model.
